Sources/WebService/AuthenticatedWebService/AuthenticatedWebService.swift

```diff
@@ -56,84 +56,83 @@ open class AuthenticatedWebService: WebService {
+        return Fail<T, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
+      }
+
+      if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
+        return self.retrySynchronizedTokenRefresh()
+          .flatMap {
+            appendTokenAndExecute(accessToken: $0)
+        }.eraseToAnyPublisher()
+      }
+      return Fail<T, Error>(error: error).eraseToAnyPublisher()
     }
-    
-    urlRequest.setValue(self.configuration.authorizationHeaderScheme.rawValue + accessToken, forHTTPHeaderField: "Authorization")
-    
-    return super.execute(urlRequest: urlRequest)
-      .catch { [weak self] error -> AnyPublisher<T, Error> in
-        guard let self = self else {
-          return Fail<T, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
-        }
-
-        if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
-          self.retrySynchronizedTokenRefresh()
-
-          return self.execute(urlRequest: urlRequest)
-            .delay(for: 0.2, scheduler: self.authenticationQueue)
-            .eraseToAnyPublisher()
-        }
-        return Fail<T, Error>(error: error).eraseToAnyPublisher()
-      }.eraseToAnyPublisher()
+    .receive(on: DispatchQueue.main)
+    .eraseToAnyPublisher()
   }
-  
-  
+
   override public func execute(urlRequest: URLRequest) -> AnyPublisher<Void, Error> {
     var urlRequest = urlRequest
-    var currentAccessToken: String?
 
-    authenticationQueue.sync {
-      currentAccessToken = self.tokenProvider.accessToken.value
+    func appendTokenAndExecute(accessToken: AccessToken) -> AnyPublisher<Void, Error> {
+      urlRequest.setValue(self.configuration.authorizationHeaderScheme.rawValue + accessToken, forHTTPHeaderField: "Authorization")
+      return super.execute(urlRequest: urlRequest)
+        .subscribe(on: DispatchQueue.global())
+        .eraseToAnyPublisher()
     }
 
-    guard let accessToken = currentAccessToken else {
-      return Fail<Void, Error>(error: NetworkError.unauthorized).eraseToAnyPublisher()
+    return Deferred {
+      self.tokenProvider.accessToken
+        .compactMap { $0 }
+        .setFailureType(to: Error.self)
+        .flatMap { accessToken -> AnyPublisher<Void, Error> in
+          return appendTokenAndExecute(accessToken: accessToken)
+        }.eraseToAnyPublisher()
+    }
+    .timeout(10, scheduler: DispatchQueue.main, customError: { NetworkError.timeout })
+    .catch { [weak self] error -> AnyPublisher<Void, Error> in
+      guard let self = self else {
+        return Fail<Void, Error>(error: NetworkError.unknown).eraseToAnyPublisher()
+      }
+
+      if self.configuration.refreshTriggerErrors.contains(where: { return $0.reflectedString == error.reflectedString }){
+        return self.retrySynchronizedTokenRefresh()
+          .flatMap {
+            appendTokenAndExecute(accessToken: $0)
```

Comment:
you have duplicated calls to `appendTokenAndExecute` in one function, which I think design issue

Reply:
yes, it looks weird indeed. This was a recursive function before, for the sake of readability and understandability we changed it to this. So the first appending happens when the request is executed for the first time, the second one is if it fails due to a 401. So it decorates the request and fires again. Should we go back to the recursion?